zb-db/src/main/java/io/zeebe/db/impl/rocksdb/ZeebeRocksDbFactory.java

```diff
@@ -55,84 +68,189 @@ private ZeebeRocksDbFactory(
+    final long blockCacheMemory = totalMemoryBudget / 3;
+    // flushing the memtables is done asynchronously, so there may be multiple memtables in memory,
+    // although only a single one is writable. once we have too many memtables, writes will stop.
+    // since prefix iteration is our bread n butter, we will build an additional filter for each
+    // memtable which takes a bit of memory which must be accounted for from the memtable's memory
+    final int maxConcurrentMemtableCount = 6;
+    final double memtablePrefixFilterMemory = 0.15;
+    final long memtableMemory =
+        Math.round(
+            ((totalMemoryBudget - blockCacheMemory) / (double) maxConcurrentMemtableCount)
+                * (1 - memtablePrefixFilterMemory));
+
+    final TableFormatConfig tableConfig = createTableFormatConfig(closeables, blockCacheMemory);
+
+    return columnFamilyOptions
+        // prefix seek must be fast, so allocate an extra 10% of a single memtable budget to create
+        // a filter for each memtable, allowing us to skip them if possible
+        .useFixedLengthPrefixExtractor(Long.BYTES)
+        .setMemtablePrefixBloomSizeRatio(memtablePrefixFilterMemory)
+        // memtables
+        // merge at least 3 memtables per L0 file, otherwise all memtables are flushed as individual
+        // files
+        .setMinWriteBufferNumberToMerge(Math.min(3, maxConcurrentMemtableCount))
+        .setMaxWriteBufferNumberToMaintain(maxConcurrentMemtableCount)
+        .setMaxWriteBufferNumber(maxConcurrentMemtableCount)
+        .setWriteBufferSize(memtableMemory)
+        // compaction
+        .setLevelCompactionDynamicLevelBytes(true)
+        .setCompactionPriority(CompactionPriority.OldestSmallestSeqFirst)
+        .setCompactionStyle(CompactionStyle.LEVEL)
+        // L-0 means immediately flushed memtables
+        .setLevel0FileNumCompactionTrigger(maxConcurrentMemtableCount)
+        .setLevel0SlowdownWritesTrigger(
+            maxConcurrentMemtableCount + (maxConcurrentMemtableCount / 2))
+        .setLevel0StopWritesTrigger(maxConcurrentMemtableCount * 2)
+        // configure 4 levels: L1 = 32mb, L2 = 320mb, L3 = 3.2Gb, L4 >= 3.2Gb
+        // level 1 and 2 are uncompressed, level 3 and above are compressed using a CPU-cheap
+        // compression algo. compressed blocks are stored in the OS page cache, and uncompressed in
+        // the LRUCache created above. note L0 is always uncompressed
+        .setNumLevels(4)
+        .setMaxBytesForLevelBase(32 * 1024 * 1024L)
+        .setMaxBytesForLevelMultiplier(10)
+        .setCompressionPerLevel(
+            List.of(
+                CompressionType.NO_COMPRESSION,
+                CompressionType.NO_COMPRESSION,
+                CompressionType.LZ4_COMPRESSION,
+                CompressionType.LZ4_COMPRESSION))
+        // defines the desired SST file size (but not guaranteed, it is usually lower)
+        // L0 => 8Mb, L1 => 16Mb, L2 => 32Mb, L3 => 64Mb
```

Comment:
I guess with big state we could still get `OutOfMemoryError` from trying to replicate? I wonder how it didn't happen, since we'd load potentially 4GB of files into memory while waiting for the replication? I mean your benchmark shows it works, but I'm surprised it does :sweat_smile:

Reply:
Yes I agree this is also what I mentioned above that we still need to fix the replication stuff.

Reply:
But I still don't get how you didn't run into it - I guess we replicate fast enough that memory gets freed fast enough and we can keep replicating? :confused: